### ISSUE TITLE
Multiple @TempDirs should result in multiple directories

### ISSUE THREAD

**Issue by sewe** [junit-team/junit5#1967]
I just got bitten by the (IMHO counter-intuitive) behavior of `@TempDir`: When using it in multiple places, each injection point gets the **same** directory injected. That is, the following does not do what one would intuitively expect:

```java
void test(@TempDir File source, @TempDir File target) {
    ...
}
```

Both `source` and `target` point to the same directory!

Of course, you can work around this:

```java
void test(@TempDir File tempDir) {
    File source = new File(tempDir, "source");
    File target = new File(tempDir, "target");
    ...
}
```

But this feels unnecessarily complex for this, IMHO, rather common use case.

If, on the other hand, injecting the same temporary directory in multiple places is _really_ what is desired, one can even imagine the following API to still support that use case:

```java
@TempDir("id") Path samePath;

void test(@TempDir("id") File sameDir) {
    ...
}
```

## Deliverables

- [ ] An update to `TemporaryDirectory` which sets up (and tears down) a new temporary directory at each injection point.

**Comment by sbrannen** [comment]
Tentatively slated for 5.6 M1 for _team discussion_

**Comment by marcphilipp** [comment]
**Team Decision:** Add to general backlog to await further interest from the community.

**Comment by iamdanfox** [comment]
@christopherpanderson and I just hit this while migrating an internal project to JUnit5, it seems quite surprising behaviour to have multiple parameters point to the same directory on disk!

The old JUnit4 code used the `TemporaryFolder#newFolder` method:

```java
@Before
public void before() {
  Path dir1 = temporaryFolder.newFolder().toPath();
  Path dir2 = temporaryFolder.newFolder().toPath();
  foo = new Foo(dir1, dir2);
}
```

I _thought_ this could just be replaced by using multiple parameters, but as this issue describes both dir1 and dir2 ended up pointing to the same directory, so the test failed in weird and wonderful ways.

```java
@BeforeEach
public void beforeEach(@TempDir Path dir1, @TempDir Path dir2) { // 🌶🌶🌶
  foo = new Foo(dir1, dir2);
}
```

**Comment by tpetracca** [comment]
Just hit this as well. Working around it in a manner similar to @sewe.

**Comment by joh-ku** [comment]
@sormuras 
Thank you very much for the quick reply in #2029 – sorry, i missed this one.
What is the actual reason for not creating multiple folders?
The workaround `Files.createDirectories(temp.resolve("one"));` applies to my issue, but is IMHO not really straightforward compared to just create muliple folders as the code suggests.

**Comment by jkosh44** [comment]
Is this issue up for grabs? If so I'd like to give it a try.

**Comment by marcphilipp** [comment]
@jkosh44 Before starting work, we need to decide if/how we want to support it.

**Comment by sormuras** [comment]
> Is this issue up for grabs?

No. Its state is: "waiting for interest" -- which seem to raise lately.

fwiw, I'm still in favor to add a "generic resource manager extension" to Jupiter, like outlined here: https://github.com/sormuras/brahms#resource-manager-extension

**Comment by OrangeDog** [comment]
Counter example: which object should be passed in each parameter?

```java

public class TestClass {

    private Path one;
    public TestClass(@TempDir Path a) {
        this.one = a;
    }

    @BeforeEach
    public void setup(@TempDir Path a, @TempDir Path b) {
        // ...
    }

    @Test
    public void test(@TempDir Path b, @TempDir Path c, @TempDir Path d) {
        // ..
    }

    @AfterEach
    public void teardown(@TempDir Path b) {
        // ..
    }

}
```

**Comment by sewe** [comment]
I would argue that each `@TempDir` would create its own directory, so all seven directories would be different.

*If* you want to re-use the same directory, I would suggest some kind of ID parameter, to support this pattern:

```java
@BeforeEach
public void setUp(@TempDir("a") Path a) {
  // Create a fresh directory with ID "a".
}

@AfterEach
public void tearDown(@TempDir("a") Path a) {
  // As the directory with ID "a" already exists, injects the existing one.
}
```

In contrast, the `@AfterEach` below would create a new directory:

```java
@AfterEach
public void tearDown(@TempDir("b") Path b) {
  // As this is the first mention of ID "b", this creates a fresh directory.
}
```

So, essentially, you would have some map of IDs to directories, with "get or create" semantics. And leaving out `TempDir.value` would simply imply a fresh ID unequal to any other ID.

**Comment by OrangeDog** [comment]
@sewe that would be a breaking change to a common use-case.

**Comment by sbrannen** [comment]
See also: #2219

**Comment by visaals** [comment]
> No. Its state is: "waiting for interest" -- which seem to raise lately.

How do we raise interest other than 👍 ing the original post? I also intuitively expected multiple `@TempDir` declarations would create unique directories and got bitten by that not being the case.

**Comment by afilip1** [comment]
Just got bitten by this unexpected behavior as well. I'd be in favor of separate directories for each parameter, but I hope at least a warning gets added, since it's quite counterintuitive as is.

**Comment by sbrannen** [comment]
> How do we raise interest other than 👍 ing the original post?

That's correct: adding a 👍 to this issue's _description_ adds a vote that can be seen in the following search results: https://github.com/junit-team/junit5/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by CRogers** [comment]
Also just got bitten by this, left me very confused why my tests were failing with files already existing.

**Comment by amirmuminovic** [comment]
I've recently started doing Java and I have to admit the behaviour of @Tempdir is not what one would expect. The fact that only one temporary folder is made could be misleading and give you the wrong impression of the test execution. Could we at least provide a warning when multiple TempDirs are used in a test?

**Comment by bakmei** [comment]
I also expect the framework to pass in a different temp folder for my tests.  Unfortunately, the test runs OK because of the hindered behavior.  My argument is that if the same value is being passed to Test1.java and Test2.java, then the existing behavior is OK.  If the same value is passed to Test1.method1() and Test1.method2(), then the existing behavior also makes sense.  But if it behaves differently at the parameter level, that breaks the pattern: class level works, method-level works, and at the parameter level is broken.  It supposed to work consistently at all levels, and that's why a few people complaining here.

In addition, the annotation is for the parameters, therefore it should work for the parameters, I would consider this is an existing bug.

**Comment by garretwilson** [comment]
I just ran into this as well. Every time I use `@TempDir`, my intuition is that "I'm using this annotation because I want a new temporary directory." After all, JUnit doesn't combine all the `@Test` annotations into a single test, does it? No, each `@Test` defines a separate unit test. And each time I use Mockito's `@Mock`, it creates a separate mock for the variable I'm annotating. (The list goes on.) In the JUnit framework I would expect each `@TempDir` likewise to create a separate temporary directory.

**Comment by mklueh** [comment]
Same here, was expecting one directory per TempDir field, but they all resolve in the same directory...if one now needs multiple test directories, you have to bloat your code with custom directory initializations in the setup method, which I like to prevent.

This behaviour should also be documented in the JavaDocs in my opinion

**Comment by marcphilipp** [comment]
**Team decision:** Based on popular demand, we will add a `value` attribute to `@TempDir` with a default value to preserve backward compatibility that will be part of the key used for storing in `ExtensionContext.Store` and hence lead to different temp dirs being created for `@TempDir("a")` and `@TempDir("b")`.

**Comment by fduminy** [comment]
> See also: #2219

As already said by @sbrannen, that other issue should be taken in consideration because in some cases the 2 issues are going in opposite directions => should #2219 be closed in favor of this issue already covering that particular case ?

**Comment by Marcono1234** [comment]
> **Team decision:** Based on popular demand, we will add a `value` attribute to `@TempDir` with a default value to preserve backward compatibility

So multiple `@TempDir` parameters or fields will continue to refer to the same directory by default?
Technically `@TempDir` is still marked as "experimental", so wouldn't it be better to adjust the default behavior with the risk of breaking backward compatibility?

It appears to me that all existing code relying on multiple `@TempDir` annotated elements referring to the same directory could be written in a saner and less error-prone way:
- Multiple annotated parameters: Just use one parameter
- Parameter and field annotated: Just use field

With the current ID approach you would end up duplicating field or parameter names as ID and even then could accidentially choose conflicting IDs.

**Comment by sbrannen** [comment]
> As already said by @sbrannen, that other issue should be taken in consideration because in some cases the 2 issues are going in opposite directions => should #2219 be closed in favor of this issue already covering that particular case ?

Thanks for bringing that up. I've just slated #2219 for 5.8 M2 as well, so that we make sure to address what should be done when multiple empty `@TempDir` annotations are declared within the same scope. Those error cases might well be handled by the implementation of this issue, in which case we would then close #2219 as superseded by this issue.

**Comment by sbrannen** [comment]
> Technically `@TempDir` is still marked as "experimental", 

That's true.

> so wouldn't it be better to adjust the default behavior with the risk of breaking backward compatibility?

I think it's worth having another brainstorming discussion within the team, so thanks for raising those points.

**Comment by marcphilipp** [comment]
**Team decision:** We are still convinced the [decision above](https://github.com/junit-team/junit5/issues/1967#issuecomment-821116112) offers the most flexibility. Doing so in conjunction with #2219 will prevent unexpected behavior like mentioned in https://github.com/junit-team/junit5/issues/1967#issuecomment-821023920.

**Comment by Marcono1234** [comment]
@marcphilipp #2219 does not mention the ID concept at the moment, but I assume a test should only fail when multiple `@TempDir` parameters or fields use the same (implicit) ID, right? Otherwise the concept of introducing an ID seems pointless since there can be at most one `@TempDir` anyways.

Though then I still don't see the advantage of having an ID because:
1. You are breaking backward compability anyways since multiple `@TempDir` with the same (implicit) ID cannot exist anymore
2. If I understand @sewe's original proposal of having an ID correctly, the idea was to allow having multiple fields / parameters for the same temp dir in case someone really wants this (for whatever reason); but that won't be possible anymore with your proposed changes. It seems the main point of this issue was to allow using multiple temp dirs. I doubt that of the current 56 'thumbs up' many actually want an ID concept if it can be solved without it as well, though please prove me wrong here.

Or is #2219 only about using `@TempDir` for both `static` and non-static variables (as mentioned in your linked comment https://github.com/junit-team/junit5/issues/1967#issuecomment-821023920)?

**Comment by marcphilipp** [comment]
> I assume a test should only fail when multiple `@TempDir` parameters or fields use the same (implicit) ID, right?

Correct.

> You are breaking backward compability anyways since multiple `@TempDir` with the same (implicit) ID cannot exist anymore

With #2219 we're deliberately breaking backward compatibility for the case where `@TempDir` is used multiple times in a method signature or the fields of a test class. We don't think anyone did this to inject the same directory multiple times but assumed they were different directories. If this breaks anyone's tests, the fix should be easy: either remove the duplicate `@TempDir` parameter/field or use an id (with this change) to disambiguate it. #2219 will also prevent multiple `@TempDir("id")` parameters/fields for consistency.

> It seems the main point of this issue was to allow using multiple temp dirs. I doubt that of the current 56 'thumbs up' many actually want an ID concept if it can be solved without it as well, though please prove me wrong here.

We've discussed this in the team and decided to introduce an ID concept now and disallow multiple parameters/fields without an ID/with the same idea. That allows declaring/using multiple temp dirs and addresses the main request of this issue.

**Comment by garretwilson** [comment]
I think one of the points that @Marcono1234 was making is that it isn't clear that preserving backwards compatibility is really the central issue here.

I would imagine that most people tried to use `@TempDir` multiple times, realized it didn't work as most people would expect it to, and then abandoned multiple use, thinking, "that's odd—why would they do it like that"? I would be interested in whether you could comb through GitHub and find many uses of people using multiple `@TempDir` instances when they wanted the _same_ temporary directory.

If your team makes that decision, that's the way the cookie crumbles. But it's not why I want, and it doesn't make sense to me. And as @Marcono1234 said, I also doubt that most of the "thumbs-up" agree with your decision. Your team will make the decision it decides. But I'm not confident that if you did a poll your actual users would prefer that decision.

**Comment by marcphilipp** [comment]
When thinking about a single test method, I agree with your expectation. However, we support `@TempDir` in other places, too. For example on `@BeforeEach` method parameters:

```java
class TempDirUsage {

	@BeforeEach
	void prepareFiles(@TempDir Path a) {
		System.out.println("a = " + a);
	}

	@Test
	void test(@TempDir Path b, @TempDir Path c) {
		System.out.println("b = " + b);
		System.out.println("c = " + c);
	}
}
```

Currently, `a`, `b`, and `c` all point to the same directory. If we change semantics for `c` to now implicitly create a separate directory from `b`, what should that mean for the value of `a`? To eliminate having to guess the semantics in this case, I think it's better to require being explicit by specifying an id when multiple temp dirs are desired and forbid injecting the same temp dir twice on the same level.

**Comment by bakmei** [comment]
@marcphilipp I would think as a developer, if I want to access the same temporary directory in two separate methods, I will provide a tag (within the scope of the test class) to the directory generator and reference it later with the same tag.  e.g. @TempDir("MyTagForAllTests").

**Comment by marcphilipp** [comment]
@bakmei That's exactly how the new solution works.

**Comment by marcphilipp** [comment]
**Team decision:** After another discussion in the team (after comments like https://github.com/junit-team/junit5/issues/2219#issuecomment-857251277), we have decided to remove the ID concept and change the semantics of `@TempDir` to always create a new directory when it's declared (regardless whether it's on a field or parameter). As a migration path, we'll keep the old semantics around for one feature release and introduce a configuration parameter to enable the old behavior so that users can upgrade and change any test broken by this change in their own timeframe. The subsequent feature release will throw an exception when it encounters the configuration parameter being set.

**Comment by bakmei** [comment]
The concept of using the value attribute would solve the test cases that need to initialize the testing environment.  As a developer, I would like to test my file system API that writes and reads to the same environment in two methods, I can use the @Before notation to set up the file system by a given temporary directory.  I can reuse the environment if needed, or create multiple environments.  I believe this will give flexibility to the user if they want to use the same environment whilst this will also maintain backward compatibility as well or it will break the old tests without giving out transit to the new implementation.

Please reconsider the path to make it all creating new ones.  Why do we need static class variables?

**Comment by marcphilipp** [comment]
@bakmei They don't have to be static. If you want to use same temp dir in a `@BeforeEach` method and a `@Test` method, you'd make it an instance field:
```java
class MyTests {
	@TempDir Path tempDir;
	@BeforeEach void prepare() {
		// prepare tempDir
	}
	@Test void test() {
		// use tempDir
	}
}
```

**Comment by bakmei** [comment]
@marcphilipp That makes sense, thanks!
